### ISSUE TITLE
A blocking mode for Processing Patterns

### ISSUE THREAD

**Issue by Chromotron** [AppliedEnergistics/Applied-Energistics-2#1778]
Suggestion: add an option for Processing Patterns to be blocking.

This means: a pattern is "locked" after its ressources are sent by the interface; it only unlocks when the system receives then expected product(s). Thus the system won't use a pattern until it receives the output from a previous usage. Especially, it should not feed in the next set of ressources before it is finished.

Currently, the only option is to not feed ressources in unless the container is empty. But more complex machines are not a single block/inventory, thus leading to problems when only the first one is emptied during the process.
The most problematic ones are machines that require an exact amount of something dropped/placed into the world (ThaumCraft cauldrons or ChromatiCraft Chroma for example). I am also not aware of any other mod solving these issues (Translocators don't, for example), and even if, this sounds like something that is best done on AE2's side.

**Comment by yueh** [comment]
Blocking is really done by the interface and not patterns (or anything else involed in autocrafting).
And besides "inventory not empty" there are not that really other options which apply to all machines.

Also it is really impossible to autocraft or is just a bit more complex than "put interface down and done"?
For example I currently cannot imagine that the Thaumcraft cauldron is not automatable with just AE2. But then there is also Thaumcraft's own autocrafting, which is probably even easier to automate.
(No idea about ChromatiCraft)

**Comment by Chromotron** [comment]
I don't see a strict reason to not have this pattern-wise (except that coding it that way might be harder). Anyway, interfaces are fine, too.

It is not just "a bit more complex". The simplest thing I came up with for those cases are essentially:
- sort stuff by type (conduits, hopper, whatever)
- build a hopper counter for every single item
- build some airlock-style gating systems
- combine with a redstone clock and/or a detector if some desired itemw as produced.
And as soon as you have two recipes using a common ingredient, it gets exponentially more complex, soon reaching the point where having a single production complex for every recipe becomes simpler than that. And that's unviable for things like the magic mods I mentioned above due to there being many recipes. Many other mods have similiar things, just picked those because they are especially annoying to automate. It's not like I couldn't build those things (I planned and built entire sequoia-to-charcoal farms in vanilla).

This is not about simplifying everything down to stupidity. It is about making it at least reasonable to not require an entire complex of machines for each single product. I don't even see how blocking by itself makes those things trivial. It's not like it becomes a triviality to do some of these by that.
If you don't believe me: try automating the cauldron recipes (all of them, not just one) with a single cauldron (or maybe a small number, but not tens or hundreds) only using AE2. Then you might realise how that simply blocking feature would come in very very handy. I don't say it is impossible without, but it becomes pretty complicated if you don't want bad things to happen while still doing mroe than one crafting per several minutes.
And no cheating by only ordering products one (type) at a time; that is easy to automate.


Anyway, I find it pretty rude to accuse me of lazyness or inability (whatever it is) as you did. That feature is in no way trivializing things; it actually makes things viable with AE2 that normally could only be done with Steve's factory manager (of which I am no fan) or Computercraft/OpenComputers (which are sometimes fun to use, but using them for crafting becomes dull very fast).
I spent some hours pondering on what exactly I would want that should also follow the spirit of AE2 and not just be another magic block. I found that the simple diea of blocking would do the trick.
If you simply don't like the idea or don't want suggestions you could have just said so; the subtle accusations were uncalled for.

**Comment by yueh** [comment]
Not accusing you of anything. Just some mods are not designed to be automatable at lower tiers and want to lock this into a higher tier like Thaumcraft.

For example what can be done for the cauldron is simply: `interface in blocking mode -> chest -> import bus -> formation plane`. This will lead the autocrafting to put exactly the amounts of 1 crafting operation into the chest and wait until everything is dropped. But for whatever reason Thaumcraft does not accept the last ingredient when not dropped by a player. I tried it a couple of times and the order was always correct, but never accepted the say last glowstone for nitor. Throwing the last one manually works just fine.
So even if we could block a crafting CPU until one operation is done, it might simply not be possible.

There would be also a couple of other issues. Like we cannot track that the item was actually crafted by your setup. Just that something entered the network. If you for example request some Nitor and Alumentum in parallel and then drop the left over amount in your inventory into the network it will completely stall the crafting as it did detect some Nitor entering the system, unblock the crafting and just mess everything up.

Patterns might simply not be the right approach (or even working).
But patterns are not the only source for autocrafting, level emitters can do it too.
They can emit a redstone to initiate the crafting of their defined item, but currently it is limited to a constant signal until done or the crafting cpu moves to another task. I already considered giving them an option to pulse once for each crafting operation. Which can essentially be use like your suggested blocking mode. Of course it will require a bit more work, say to pulse a couple of export buses to export the right amount of items in the right order. Or even stagger multiple crafting cpus to break up requesting N items of X into guaranteed jobs with 1 item of X in another network.
Even with a blocking mode on patterns it can easily mess up, let us keep the example of crafting Nitor and Alumentum in parallel. If the blocking is on the level of patterns it will not prevent both crafting cpus from interfering with each other at least not without building the same safeguards currently already needed. Being able to delegate this to a seperate network with only one crafting cpu would technically act as safeguard and prevent any parallel execution of tasks which could break.

On a second thought, this should be already possible even with a constant redstone signal.
Pulsing the level emitter is probably not even needed.
I really want a reason to add something like the pulsing emitter, but each time I am pondering on it, it usually ends up not being necessary and can be already done with AE2/vanilla. So literally a violation of your own [suggestion guidelines](http://ae-mod.info/Suggestion-Guidelines/).
All I really want is some serious discussion about it, to find a good solution and let it mature a bit instead of just starting to add it and then realizing it is already possible or worst case even a bad feature.

**Comment by bookerthegeek** [comment]
The reason that you cant "fully" automate the cauldron fyi is that it checks if the player dropping the last item has learned the recipe. I assume "YOU" know the recipe cause your trying to automate it. But how in the world would an AE network learn the recipe?

**Comment by thatsIch** [comment]
FakePlayer! xD
